### PR TITLE
With frame-pointers, exception handlers should restore RBP

### PR DESCRIPTION
**Pull request by fabbing**
# Frame-pointers in a nutshell
Frame-pointers can be used to walk the stack by tools ranging from debuggers to profilers (ie perf, asan, tsan).
On x86-64, the `rbp` register is reserved to reference the base address of the current stack frame (the start of the function stack frame) while `rsp` points to the top of the current stack frame. The first word of a function stack frame is used to save the value of the `rbp` register at function entry (the stack base address of the *caller* function).
This effectively creates a linked list, starting from the `rbp` register, to the begin of each function frame on the stack.

# Exception handlers don't restore `rbp`
I have noticed that exception handlers (`ignore (handler i)` inside the `catcher` function in the example) do not restore the `rbp` register when entered. `rbp` is potentially left pointing inside the stack of frame of a nested function called when evaluating the try expression (`raiser` in the example).
If the handler calls a new function using stack space (`handler` in the example), the stack position pointed to by `rbp` is overwritten and the backtrace becomes unusable.

# Illustration with one example
By placing breakpoints at specific locations it is possible to observe the `rbp` based backtrace becoming corrupted.

## Source
```ocaml
exception FortyTwo

let [@inline never] handler i =
  Printf.printf "handler %d\n" i;
  i + 1

let [@inline never] raiser i =
  Printf.printf "raiser %d\n" i;
  match i with
  | 42 -> raise FortyTwo
  | _ -> i

let [@inline never] catcher i =
  Printf.printf "catcher %d\n" i;
  try raiser i with
  | FortyTwo -> ignore (handler i);
  i

let () =
  ignore (catcher 42)
```

## Breakpoint in `camlRbp__raiser_307` before raising
At this point we are in the `raiser` function. We have already used some stack space for local variables and will soon raise an exception.

### Backtrace with DWARF
```
#0  camlRbp__raiser_307 () at rbp.ml:10
#1  0x000055caa78f9fe2 in camlRbp__catcher_310 () at rbp.ml:15
#2  0x000055caa78fa082 in camlRbp__entry () at rbp.ml:20
#3  0x000055caa78f78f9 in caml_program ()
#4  0x000055caa793bacd in caml_start_program ()
#5  0x000055caa793c31c in caml_startup_common (argv=0x7ffd6c8c1fb8, pooling=<optimized out>, pooling@entry=0) at startup_nat.c:160
#6  0x000055caa793c39b in caml_startup_exn (argv=<optimized out>) at startup_nat.c:167
#7  caml_startup (argv=<optimized out>) at startup_nat.c:172
#8  caml_main (argv=<optimized out>) at startup_nat.c:179
#9  0x000055caa78f76dc in main (argc=<optimized out>, argv=<optimized out>) at main.c:37
```

### Stack layout
| Address | Content | Comment | Pointed by |
| --- | --- | --- | --- |
| 0x7ffd6c8c1d88 | 0x000055caa78f9fe2 | return address | |
| 0x7ffd6c8c1d80 | 0x00007ffd6c8c1db0 | saved rbp | <- rbp |
| 0x7ffd6c8c1d78 | 0x0000000000000055 | local variable, Val(42) | |
| 0x7ffd6c8c1d70 | 0x0000000000000055 | local variable, Val(42) | <- rsp |

## Breakpoint in `camlRbp__handler_268`
The exception was raised and caught by the handler of the `try with` in `catcher`.
But the handler has called a new function `handler` which used stack space and overwrote the memory address still pointed by our not-updated `rbp` register.

### Backtrace with DWARF
```
#0  camlRbp__handler_268 () at rbp.ml:5
#1  0x000055caa78fa005 in camlRbp__catcher_310 () at rbp.ml:16
#2  0x000055caa78fa082 in camlRbp__entry () at rbp.ml:20
#3  0x000055caa78f78f9 in caml_program ()
#4  0x000055caa793bacd in caml_start_program ()
#5  0x000055caa793c31c in caml_startup_common (argv=0x7ffd6c8c1fb8, pooling=<optimized out>, pooling@entry=0) at startup_nat.c:160
#6  0x000055caa793c39b in caml_startup_exn (argv=<optimized out>) at startup_nat.c:167
#7  caml_startup (argv=<optimized out>) at startup_nat.c:172
#8  caml_main (argv=<optimized out>) at startup_nat.c:179
#9  0x000055caa78f76dc in main (argc=<optimized out>, argv=<optimized out>) at main.c:37
```

### `rbp` based backtrace
This `rbp` based backtrace is created by blindly following the `rbp` linked list to walk the stack.
Since one word of the linked list has been overwritten by a local variable, it is no longer walkable.
```
current rip is (void *) 0x55caa78f9f0c <camlRbp__handler_268+60>
saved rip @ 0x0000006c8c1d98 is (void *) 0x55caa78fa005 <camlRbp__catcher_310+117>
saved rbp @ 0x0000006c8c1d90 is 0x0000006c8c1d80
saved rip @ 0x0000006c8c1d88 is (void *) 0x55
saved rbp @ 0x0000006c8c1d80 is 0x00000000000055
saved rip @ 0x0000000000005d is Cannot access memory at address 0x5d
```

### Stack layout
Notice that the saved `rbp` in `camlRbp__handler_268` (at 0x7ffd6c8c1d90) which should point inside `camlRbp__catcher_310` (at 0x7ffd6c8c1db0) has the previous, non-updated, value of 0x00007ffd6c8c1d80 which points *inside* our own stack frame and this location has just been overwritten by local variables.

| Address | Content | Comment | Pointed by |
| --- | --- | --- | --- |
| camlRbp__catcher_310 |
| 0x7ffd6c8c1db8 | 0x000055caa78fa082 | return address in camlRbp__entry+98 | |
| 0x7ffd6c8c1db0 | 0x00007ffd6c8c1dc0 | saved rbp for camlRbp__entry | |
| 0x7ffd6c8c1da8 | 0x0000000000000055 | local variable, Val(42) | |
| 0x7ffd6c8c1da0 | 0x0000000000000055 | local variable, Val(42) | |
| camlRbp__handler_268 |
| 0x7ffd6c8c1d98 | 0x000055caa78fa005 | return address in camlRbp__catcher_310 | |
| 0x7ffd6c8c1d90 | 0x00007ffd6c8c1d80 | saved *wrong* rbp | <- rbp |
| 0x7ffd6c8c1d88 | 0x0000000000000055 | local variable, Val(42) | |
| 0x7ffd6c8c1d80 | 0x0000000000000055 | local variable, Val(42) | <- rsp |

Once `camlRbp__catcher_310` completes, the epilogue restores `rbp` to a reasonable value and the problem disappears.

# Proposed solution

```ocaml
   | Lentertrap ->
      if fp then
        let delta = (frame_size env) - 16 (* retaddr + rbp *) in
        I.lea (mem64 NONE delta RSP) rbp;
```

The proposed solution is much shorter than all the verbiage above:
Just before entering the exception handler, its trap is "pop-ed" from the stack (https://github.com/ocaml/ocaml/blob/b53c79e59bebc4ffe3b114f394acc10a886143bf/runtime/amd64.S#L528) and `rsp` is shifted accordingly.
Since the compiler knows the size of the frame, we use the `Lentertrap` instruction to perform some advanced mathematics (a subtraction) and compute the expected `rbp` value from `rsp`.

# Acknowledgement
Thanks to @ctk21 and @mshinwell for confirming with me the existence of this problem and to @stedolan for pointing out almost exactly this solution.

### CODE REVIEW

File: asmcomp/amd64/emit.mlp [hunk 1]
```diff
@@ -804,7 +804,9 @@ let emit_instr env fallthrough i =
       done;
       emit_named_text_section env.f.fun_name
   | Lentertrap ->
-      ()
+      if fp then
+        let delta = (frame_size env) - 16 (* retaddr + rbp *) in
+        I.lea (mem64 NONE delta RSP) rbp;
```
Comment (xavierleroy) [inline asmcomp/amd64/emit.mlp]: It's hard to guess the relative precedence of `if`, `let` and semicolon, but easy to guess that of function application and subtraction.  Suggested style change:
```suggestion
      if fp then begin
        let delta = frame_size env - 16 (* retaddr + rbp *) in
        I.lea (mem64 NONE delta RSP) rbp
      end;
```

**Review by xavierleroy — APPROVED** [review]
Well spotted.  The proposed fix looks good to me.

**Comment by gasche** [comment]
@fabbing could you add a Changes entry in the "Bug fixes" section?

**Comment by gasche** [comment]
@Octachron this PR targets 4.14 (submitter choice), and it is ripe for merging. Are you ok with the 4.14 destination?

**Comment by Octachron** [comment]
The bug is not really recent, isn't it?

On the one hand, the fix looks like it might have not so straightforward failure modes.

On the other hand, it only matters when the `frame-pointer` configuration option is enabled which is not heavily tested during the alpha and beta releases.

Overall, my balance is tilted towards the "ok for 4.14.0" side.

**Comment by gasche** [comment]
Yes, I agree that it looks a bit tricky to go in 4.14 at the last minute, but you decide :-)
(On the "let's go" side: the PR looks very carefully prepared, not to speak of our outstanding reviewer.)

**Comment by fabbing** [comment]
I'm trying to add a test to the testsuite that would exercise this change but it's trickier than expected, so I will open a new PR for the test(s), not to prevent this one to be merged.

**Comment by xavierleroy** [comment]
Don't go overboard with the test: it's unlikely that this particular bug will be reintroduced later, given how simple and natural the fix is.

**Comment by xavierleroy** [comment]
> On the one hand, the fix looks like it might have not so straightforward failure modes.

The fix is very safe: in frame-pointer mode, rbp is not used by the generated code (it could contain random values and the code would still compute the right results, only backtraces and profiles would be broken).  Plus, rbp will be restored to a presumably-correct value at the next function return.  (That's why the bug went unnoticed for so long: rbp has the wrong value for a short duration.)  So, the risk of breakage is low.

**Comment by Octachron** [comment]
Thanks for the contextual information. Let's go with the merge to 4.14.0 then.

**Comment by gasche** [comment]
Merged, and cherry-picked in trunk as 9a32caf8b7e9ee971dbb34703d5a07d3e20006cd.

**Pull request merged**